### PR TITLE
Rewrite udev-test.pl in Python

### PR DESCRIPTION
**Pull request by keszybz**
@DaanDeMeyer you asked for this.

### CODE REVIEW

**Comment by DaanDeMeyer** [comment]
Not removing the perl one yet?

**Comment by keszybz** [comment]
I think that in the tests, the perl and python versions were running concurrently and interfering with each other. I dropped the perl version now, and also converted another script that we had. What now remains is a bit of perl use in tests. We could get rid of this too without too much trouble, and then we could drop the perl dependency altogether.

**Comment by keszybz** [comment]
I think that in the tests, the perl and python versions were running concurrently and interfering with each other. I dropped the perl version now, and also converted another script that we had. What now remains is a bit of perl use in tests. We could get rid of this too without too much trouble, and then we could drop the perl dependency altogether.

**Comment by DaanDeMeyer** [comment]
So CentOS CI Arch is failing because python3 does not exist on Arch Linux, they only have python. Not sure how we haven't come across this before.

**Comment by mrc0mmand** [comment]
> So CentOS CI Arch is failing because python3 does not exist on Arch Linux, they only have python. Not sure how we haven't come across this before.

Nah, it does exist:

```
# pacman -Qo /usr/bin/python3
/usr/bin/python3 is owned by python 3.11.3-1
```

we just don't install it into the test image (I don't think we install python at all into the test images).

**Comment by mrc0mmand** [comment]
> > ```
> > /usr/bin/python3 is owned by python 3.11.3-1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > we just don't install it into the test image (I don't think we install python at all into the test images).
> 
> OK, so what should I do here?

Heh, interestingly enough we don't install perl either, but the udev test didn't fail since TEST-02 runs only files prefixed with `test-` and the original perl one was named `udev-test.pl`. So I guess the easiest "solution" for now would be:

```diff
diff --git a/test/units/testsuite-02.sh b/test/units/testsuite-02.sh
index 61f6d06397..20308748d8 100755
--- a/test/units/testsuite-02.sh
+++ b/test/units/testsuite-02.sh
@@ -6,7 +6,7 @@ set -o pipefail
 NPROC=$(nproc)
 MAX_QUEUE_SIZE=${NPROC:-2}
 TESTS_GLOB=${TESTS_GLOB:-test-*}
-mapfile -t TEST_LIST < <(find /usr/lib/systemd/tests/unit-tests/ -maxdepth 1 -type f -name "${TESTS_GLOB}")
+mapfile -t TEST_LIST < <(find /usr/lib/systemd/tests/unit-tests/ -maxdepth 1 -type f -name "${TESTS_GLOB}" ! -name "test-udev.py")
 
 # reset state
 rm -fv /failed-tests /skipped-tests /skipped
```

**Comment by DaanDeMeyer** [comment]
Can you drop the perl installation from mkosi and the build_test/unit_tests workflows as well?

File: README [hunk 1]
```diff
@@ -206,17 +205,17 @@ REQUIREMENTS:
         gperf
         docbook-xsl (optional, required for documentation)
         xsltproc    (optional, required for documentation)
+        python >= 3.7 (required by meson too, >= 3.9 is required for ukify)
         python-jinja2
-        python-pefile
+        python-pefile (option, required for ukify)
```
Comment (bluca) [inline README]: option -> optional

**Comment by keszybz** [comment]
Wow, it seems the CI passed. ("Lint Code Base" is complaining about code that wasn't changed in this PR.)  Merge?

**Comment by DaanDeMeyer** [comment]
@keszybz Maybe drop perl dependency from mkosi and other CI stuff?

**Comment by keszybz** [comment]
> @keszybz Maybe drop perl dependency from mkosi and other CI stuff?

Done now. I wanted to do it later in a separate step, but if you think it's better to do now, then let's try that.

File: src/test/test-udev.c [hunk 2]
```diff
@@ -123,6 +125,13 @@ static int run(int argc, char *argv[]) {
         action = argv[1];
         devpath = argv[2];
 
+        if (argv[3]) {
+                r = safe_atou(argv[3], &us);
```
Comment (yuwata) [inline src/test/test-udev.c:130]: `parse_sec()` is better??

**Review by DaanDeMeyer — APPROVED** [review]

**Comment by keszybz** [comment]
Updated again:
- `parse_usec` per request
- the change in `tools/oss-fuzz.sh` was borked and the CI caught it

**Comment by bluca** [comment]
```
14:47:59 ../build/test/test-udev.py::test_udev[TAG refuses invalid string] PASSED [ 83%]
14:47:59 ../build/test/test-udev.py::test_udev[IMPORT parent test] 
14:47:59 ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
14:47:59 
14:47:59 
14:47:59 Summary of Failures:
14:47:59 
14:47:59 1507/1507 systemd / test-udev                                                                                                                     TIMEOUT         540.39s   killed by signal 15 SIGTERM
```

**Comment by yuwata** [comment]
Hmm? Still timeouts...
```
../build/test/test-udev.py::test_udev[TAG refuses invalid string] PASSED [ 83%]
../build/test/test-udev.py::test_udev[IMPORT parent test] 
â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•


Summary of Failures:

1538/1538 systemd / test-udev                                                                                                                     TIMEOUT        1080.35s   killed by signal 15 SIGTERM
```

File: test/udev-test.py [hunk 3]
```diff
@@ -0,0 +1,2408 @@
...
+        "IMPORT parent test",
+        Device(
+            "/devices/pci0000:00/0000:00:1f.2/host0/target0:0:0/0:0:0:0/block/sda",
+            exp_links       = ["parent"],
+        ),
+        Device(
+            "/devices/pci0000:00/0000:00:1f.2/host0/target0:0:0/0:0:0:0/block/sda/sda1",
+            exp_links       = ["parentenv-parent_right"],
+        ),
+        delay = 500000,  # Serialized! We need to sleep here after adding sda
```
Comment (yuwata) [inline test/udev-test.py]: Oh, now we parse delay in sec.

File: test/udev-test.py [hunk 4]
```diff
@@ -0,0 +1,2408 @@
...
+            "/devices/pci0000:00/0000:00:1f.2/host0/target0:0:0/0:0:0:0/block/sda/sda9",
+            not_exp_links   = ["partition"],
+            exp_links       = ["part-9"],
+        ),
+        Device(
+            "/devices/pci0000:00/0000:00:1f.2/host0/target0:0:0/0:0:0:0/block/sda/sda10",
+            not_exp_links   = ["partition"],
+            exp_links       = ["part-10"],
+        ),
+        delay = 10000,
```
Comment (yuwata) [inline test/udev-test.py]: Also here.

File: test/udev-test.py [hunk 5]
```diff
@@ -0,0 +1,2408 @@
...
+                     not_exp_links=not_exp)
+
+
+@dataclasses.dataclass
+class Rules:
+    desc: str
+    devices: list[Device]
+    rules: str
+    repeat: int = 1
+    delay: Optional[float] = None
```
Comment (yuwata) [inline test/udev-test.py]: float -> string or unsigned?

File: test/udev-test.py [hunk 6]
```diff
@@ -0,0 +1,2408 @@
...
+        """),
+]
+
+def fork_and_run_udev(action: str, rules: Rules) -> None:
+    kinder = []
+    for k, device in enumerate(rules.devices):
+        # TODO: valgrind/gdb/strace
+        cmd = [UDEV_BIN, action, device.devpath]
+        if rules.delay:
+            cmd += [f'{k * rules.delay}']
```
Comment (yuwata) [inline test/udev-test.py]: Why `k` is multiplied to delay?

Comment (keszybz) [reply test/udev-test.py]: That's what the original code was doing. AFAUI, it wants to start the jobs staggered.

File: test/udev-test.py (hunk 5 above)
Comment (keszybz) [reply test/udev-test.py]: There is no unsigned in python. I changed it to `int` now.

File: test/udev-test.py (hunk 3 above)
Comment (keszybz) [reply test/udev-test.py]: Ah, right. I undid the change to `parse_usec()`. Using seconds just makes this all more complicated for no gain.

**Comment by keszybz** [comment]
Note: I dropped the bump of the timeout. The timeout was probably caused by the mixup between µs and s. The test should now pass with the old timeout.

**Pull request merged**